projects/chaos/src/main.c

```diff
@@ -1,55 +1,113 @@
-// Test program for STM32F072RB discovery boards - attempts to connect to the onboard MEMS
-// gyroscope over SPI. Blinks the blue LED on success or the red LED on fail.
 #include <stdbool.h>
+#include <stddef.h>
 #include <stdint.h>
+#include <stdlib.h>
 
+#include "adc.h"
+#include "can.h"
+#include "can_msg_defs.h"
+#include "can_unpack.h"
+#include "chaos_config.h"
+#include "chaos_events.h"
+#include "event_queue.h"
 #include "gpio.h"
+#include "gpio_it.h"
+#include "gpio_seq.h"
+#include "interrupt.h"
 #include "log.h"
-#include "spi.h"
+#include "power_path.h"
+#include "soft_timer.h"
+#include "status.h"
 
-#define GYRO_ID 0xD4
+#define CHAOS_NUM_RX_HANDLERS 1
+#define CHAOS_GPIO_OPEN_DELAY_US 1000
 
-void gyro_cmd(bool read, bool autoincrement, uint8_t addr, uint8_t *data) {
-  uint8_t packet[] = { (read & 0x01) << 7 | (autoincrement & 0x01) << 6 | (addr & 0x3F),
-                       (read) ? 0 : *data };
+static CANStorage s_can_storage;
+static CANRxHandler s_rx_handlers[CHAOS_NUM_RX_HANDLERS];
 
-  spi_exchange(SPI_PORT_2, packet, (size_t)(2 - read), data, read);
+static StatusCode prv_handle_uvov(const CANMessage *msg, void *context, CANAckStatus *ack_reply) {
+  (void)ack_reply;
+  (void)context;
+  uint8_t dcdc_ov = 0;
+  uint8_t dcdc_uv = 0;
+  uint8_t aux_ov = 0;
+  uint8_t aux_uv = 0;
+  CAN_UNPACK_OVUV_DCDC_AUX(msg, &dcdc_ov, &dcdc_uv, &aux_ov, &aux_uv);
+  LOG_DEBUG("Caught UVOV Message | AUX - UV: %d OV: %d | DCDC - UV: %d OV: %d\n", aux_uv, aux_ov,
+            dcdc_uv, dcdc_ov);
+  return STATUS_CODE_OK;
 }
 
+// Temporary debug main for testing.
 int main(void) {
+  event_queue_init();
+  interrupt_init();
+  soft_timer_init();
   gpio_init();
+  gpio_it_init();
+  adc_init(ADC_MODE_SINGLE);
 
-  SPISettings spi_settings = {
-    .baudrate = 1500000,
-    .mode = SPI_MODE_0,
-    .mosi = { GPIO_PORT_B, 15 },
-    .miso = { GPIO_PORT_B, 14 },
-    .sclk = { GPIO_PORT_B, 13 },
-    .cs = { GPIO_PORT_C, 0 },
+  CANSettings can_settings = {
+    .device_id = SYSTEM_CAN_DEVICE_CHAOS,
+    .bitrate = CAN_HW_BITRATE_125KBPS,
+    .rx_event = CHAOS_EVENT_CAN_RX,
+    .tx_event = CHAOS_EVENT_CAN_TX,
+    .fault_event = CHAOS_EVENT_CAN_FAULT,
+    .tx = { GPIO_PORT_A, 12 },
+    .rx = { GPIO_PORT_A, 11 },
+    .loopback = true,
```

Comment:
Probably want to switch this :P

Reply:
This is just a debugging script. I'll remove it in the real main.c which I'll put up shortly.